### PR TITLE
Dirty version of interning StructIdentifier

### PR DESCRIPTION
**Pull request by ziaptos**
Added a `instance_uid` that maps an instance to a unique identifier (we use smart pointers for the underlying uid

### Test Plan
Separate tests to be provided

### CODE REVIEW

File: third_party/move/move-vm/types/src/loaded_data/instance_uid.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+}
+impl<T: Eq + Hash> Eq for ArcKeyByPtr<T> {}
+impl<T: Eq + Hash> Ord for ArcKeyByPtr<T> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        Arc::as_ptr(&self.key).cmp(&Arc::as_ptr(&other.key))
+    }
+}
+impl<T: Eq + Hash> PartialOrd for ArcKeyByPtr<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
```
Comment (brmataptos) [inline third_party/move/move-vm/types/src/loaded_data/instance_uid.rs:50]: Why not use `as_ptr` here, too?  For BTree* that would be a big win.

File: third_party/move/move-vm/types/src/loaded_data/instance_uid.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    hash::{Hash, Hasher},
+    sync::{Arc, Weak},
+};
+
+// TODO: add tests.
+// TODO: break into files.
+// TODO: micro benchmarks.
+
+pub struct GlobalCache<'a, T: Eq + Hash> {
+    instance_to_compressed: DashMap<Arc<T>, Weak<InstanceUIDInner<'a, T>>>,
```
Comment (brmataptos) [inline third_party/move/move-vm/types/src/loaded_data/instance_uid.rs:20]: "compressed" is wrong here.  intern_table?

File: third_party/move/move-vm/types/src/loaded_data/instance_uid.rs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+pub struct InstanceUIDInner<'a, T: Eq + Hash> {
+    key: ArcKeyByPtr<T>,
+    #[derivative(
+        Debug = "ignore",
+        Hash = "ignore",
+        PartialEq = "ignore",
+        PartialOrd = "ignore",
+        Ord = "ignore"
+    )]
+    map: &'a DashMap<Arc<T>, Weak<InstanceUIDInner<'a, T>>>,
```
Comment (brmataptos) [inline third_party/move/move-vm/types/src/loaded_data/instance_uid.rs:70]: Why do you need a reference to the map here?

File: third_party/move/move-vm/types/src/loaded_data/instance_uid.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                    if let Some(compressed) = Weak::upgrade(entry.get()) {
+                        return compressed;
+                    }
+                },
+                Vacant(entry) => {
+                    let inner = InstanceUIDInner {
+                        key: ArcKeyByPtr::new(instance_arc),
+                        map: &self.instance_to_compressed,
+                    };
+                    let ret = Arc::new(inner);
```
Comment (brmataptos) [inline third_party/move/move-vm/types/src/loaded_data/instance_uid.rs:114]: Why 2 layers of Arcs here?  I would expect that the one in ArcKeyByPtr would be all you need.

File: third_party/move/move-vm/types/src/loaded_data/instance_uid.rs (hunk 2 above)
Comment (brmataptos) [reply third_party/move/move-vm/types/src/loaded_data/instance_uid.rs:20]: Also, as mentioned below: why is the value an InstanceUIDInner<...>?  Why not just a Weak<T> which is downgraded Arc<T>?  (Actually, you probably want another wrapper struct like ArcKeyByPtr to enforce handling).

File: third_party/move/move-vm/types/src/loaded_data/instance_uid.rs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        Some(self.cmp(other))
+    }
+}
+impl<T: Eq + Hash> Hash for ArcKeyByPtr<T> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        Arc::as_ptr(&self.key).hash(state);
+    }
+}
+
+#[derive(Derivative)]
```
Comment (brmataptos) [inline third_party/move/move-vm/types/src/loaded_data/instance_uid.rs:59]: What value is InstanceUIDInner providing?

File: third_party/move/move-vm/types/src/loaded_data/instance_uid.rs (hunk 2 above)
Comment (ziaptos) [reply third_party/move/move-vm/types/src/loaded_data/instance_uid.rs:20]: Yes- "compressed is wrong". I don't really like intern_table either. It is cache in a way that keeps track only of "alive" instances.

File: third_party/move/move-vm/types/src/loaded_data/instance_uid.rs (hunk 5 above)
Comment (ziaptos) [reply third_party/move/move-vm/types/src/loaded_data/instance_uid.rs:59]: It's drop method ensures we remove the instance from the table of alive instances

File: third_party/move/move-vm/types/src/loaded_data/instance_uid.rs (hunk 3 above)
Comment (ziaptos) [reply third_party/move/move-vm/types/src/loaded_data/instance_uid.rs:70]: To remove it from the map when the last instance drops

File: third_party/move/move-vm/types/src/loaded_data/instance_uid.rs (hunk 4 above)
Comment (ziaptos) [reply third_party/move/move-vm/types/src/loaded_data/instance_uid.rs:114]: The inner Arc can be actually avoided, tho will require some "unsafe" code.

**Review by ziaptos — COMMENTED** [review]
I'm still working on a large refactor of this. The notion doc can answer bunch of questions about the design chouces.

File: third_party/move/move-vm/types/src/loaded_data/instance_uid.rs (hunk 1 above)
Comment (ziaptos) [reply third_party/move/move-vm/types/src/loaded_data/instance_uid.rs:40]: We need to collectively decide whether we want to break the orderings in BTree*.

**Comment by brmataptos** [comment]
UID in the name is poor, as it suggests a pointer to a pure identifier (or, worse, a string UUID).  As it's really an Arc underneath, users should avoid Copy if they can and prefer to use & for local references.  Your name should be more intuitive.

Do we handle correctly the case where an entry is being dropped just as another thread reads it from the map?

File: third_party/move/move-vm/runtime/src/loader/mod.rs [hunk 6]
```diff
@@ -1736,8 +1739,8 @@ impl StructInfoCache {
 
 #[derive(Clone)]
 pub(crate) struct TypeCache {
-    structs: HashMap<StructIdentifier, HashMap<Vec<Type>, StructInfoCache>>,
-    depth_formula: HashMap<StructIdentifier, DepthFormula>,
+    structs: HashMap<StructIdentifierUID, HashMap<Vec<Type>, StructInfoCache>>,
```
Comment (gelash) [inline third_party/move/move-vm/runtime/src/loader/mod.rs:1742]: @ziaptos did you say BTreeMap worked better here?

**Pull request closed without merging**